File: src/main/java/gg/mineral/practice/util/messages/StringUtil.java
package gg.mineral.practice.util.messages;

import org.bukkit.ChatColor;

public class StringUtil {
    public static String toNiceString(String string) {
        string = ChatColor.stripColor(string).replace('_', ' ').toLowerCase();
        final StringBuilder sb = new StringBuilder();
        for (int i = 0; i < string.toCharArray().length; ++i) {
            char c = string.toCharArray()[i];
            if (i > 0) {
                final char prev = string.toCharArray()[i - 1];
                if ((prev == ' ' || prev == '[' || prev == '(') && (i == string.toCharArray().length - 1 || c != 'x'
                        || !Character.isDigit(string.toCharArray()[i + 1]))) {
                    c = Character.toUpperCase(c);
                }
            } else if (c != 'x' || !Character.isDigit(string.toCharArray()[i + 1])) {
                c = Character.toUpperCase(c);
            }
            sb.append(c);
        }
        return sb.toString();
    }
}

File: src/main/kotlin/gg/mineral/practice/util/messages/StringUtil.kt
package gg.mineral.practice.util.messages

import org.bukkit.ChatColor
import java.util.*

object StringUtil {
    fun toNiceString(string: String): String {
        val newString = ChatColor.stripColor(string)
            .replace('_', ' ')
            .lowercase(Locale.getDefault())
        val chars = newString.toCharArray()
        val length = chars.size
        val sb = StringBuilder()

        for (i in 0 until length) {
            var c = chars[i]
            if (i > 0) {
                val prev = chars[i - 1]
                val spaceOrBracket = prev in setOf(' ', '[', '(')
                if (spaceOrBracket) {
                    val isLast = i == length - 1
                    val nextIsDigit = if (i < length - 1) Character.isDigit(chars[i + 1]) else false
                    if (isLast || c != 'x' || !nextIsDigit) {
                        c = c.uppercaseChar()
                    }
                }
            } else {
                val isX = c == 'x'
                val nextIsDigit = if (length - 1 > 0) Character.isDigit(chars[1]) else false
                if (!isX || !nextIsDigit) c = c.uppercaseChar()
            }

            sb.append(c)
        }

        return sb.toString()
    }


}
